lib/AST/UnqualifiedLookup.cpp

```diff
@@ -430,11 +399,29 @@ ValueDecl *UnqualifiedLookupFactory::lookupBaseDecl(const DeclContext *baseDC) c
   // In these cases, using the Swift 6 lookup behavior doesn't affect
   // how the body is type-checked, so it can be used in Swift 5 mode
   // without breaking source compatibility for non-escaping closures.
-  if (!Ctx.LangOpts.isSwiftVersionAtLeast(6) &&
-      !implicitSelfReferenceIsUnwrapped(selfDecl, closureExpr)) {
+  if (capturesSelfWeakly && Ctx.LangOpts.isSwiftVersionAtLeast(6) &&
+      !implicitSelfReferenceIsUnwrapped(selfDecl)) {
     return nullptr;
   }
 
+  // Closures are only allowed to rebind self in specific circumstances:
+  //  1. In a capture list with an explicit capture.
+  //  2. In a `guard let self = self` / `if let self = self` condition
+  //     in a closure that captures self weakly.
+  //
+  // These rebindings can be done by any parent closure, and apply
+  // to all nested closures. We only need to check these structural
+  // requirements loosely here -- more extensive validation happens
+  // later in MiscDiagnostics::diagnoseImplicitSelfUseInClosure.
+  //
+  // Other types of rebindings, like an arbitrary "let `self` = foo",
+  // are never allowed to rebind self.
+  if (auto var = dyn_cast<VarDecl>(selfDecl)) {
+    if (!(var->isCaptureList() || var->getParentPatternStmt())) {
+      return nullptr;
```

Comment:
One interesting question is also how we ought to handle things like this:

```swift
struct S {
  func bar() {
    let fn = { [self = S()] in
      bar()
    }
  }
}
```

Today this ignores the bound `self` and uses the `self` of the method. With your change, we will switch to the capture list, but not diagnose because it's not a class type. I wonder if the validation of the `self` rebinding ought to be its own diagnostic that applies generally to any implicit use of `self` that isn't a rebinding of the `self` of the context, which we could stage in via a warning.

Reply:
> One interesting question is also how we ought to handle things like this:
> 
> ```swift
> struct S {
>   func bar() {
>     let fn = { [self = S()] in
>       bar()
>     }
>   }
> }
> ```
> 

Thanks for thinking of this, I hadn't considered that case before. I added an error for this in Swift 6 mode, staged in with a warning in Swift 5 mode.

> Just realized we accept arbitrary non-VarDecl results here, e.g this becomes valid:
> 
> ```swift
> class C {
>   func bar() {
>     struct `self` {
>       static func bar() {}
>     }
>     let fn = {
>       bar()
>     }
>   }
> }
> ```
> 
> Should we be rejecting non-VarDecl results here?

Good thinking. I tested this and is was rejected as expected, but I went ahead and also rejected non-VarDecls here to be defensive, and added tests for this sort of case.